### ISSUE TITLE
Filter specific pipelines for metric CI-pipeline duration

### ISSUE THREAD

**Issue by Sebastiaan127001** [ICTU/quality-time#10426]
**Is your feature request related to a problem? Please describe.**
When using the metric CI-pipeline duration with GitLab as source, and configuring scheduled pipeline runs, GitLab may return multiple pipelines and it is not possible to further select a specific pipeline.

**Describe the solution you'd like**
A possibility to filter (scheduled) pipelines by description.
Also include the matched pipelines in the measurement entities.

**Additional context**
Unfortunately, the pipelines endpoints do not include the schedule id in the responses. So, to know which pipeline was triggered by which schedule, we need to retrieve the schedules via the [schedules endpoint](https://docs.gitlab.com/ee/api/pipeline_schedules.html) and then get the pipelines scheduled by the schedule via the [ pipelines triggered by a pipeline schedule endpoint](https://docs.gitlab.com/ee/api/pipeline_schedules.html#get-all-pipelines-triggered-by-a-pipeline-schedule). 

The scheduled pipelines response includes the pipeline schedule description, so users can use that to filter schedules. No need to use the schedule id.

**Comment by Sebastiaan127001** [comment]
this is for project C

**Comment by fniessink** [comment]
It's already possible to filter pipelines by trigger. One of the triggers is `schedule`. Can't you achieve the desired effect by setting the value of the "Pipeline triggers to include" parameter to all triggers except `schedule`?

**Comment by Sebastiaan127001** [comment]
But it only the scheduled pipelines I'm interested in.

**Comment by javanegmond** [comment]
The information of the last execution of a scheduled pipeline is available through a GET on the `pipeline_schedules/:id` endpoint of the GitLab API:
https://docs.gitlab.com/ee/api/pipeline_schedules.html#get-a-single-pipeline-schedule

Does this help?

**Comment by fniessink** [comment]
> The information of the last execution of a scheduled pipeline is available through a GET on the `pipeline_schedules/:id` endpoint of the GitLab API: https://docs.gitlab.com/ee/api/pipeline_schedules.html#get-a-single-pipeline-schedule
> 
> Does this help?

Yeah, thanks. It seems that to be able to filter by pipeline schedule description, Quality-time will have to:
1. Get all pipeline schedules and descriptions via `GET /projects/:id/pipeline_schedules`
2. Get the id of the last run via `GET /projects/:id/pipeline_schedules/:pipeline_schedule_id` (repeated call for each scheduled pipeline)
3. Get all pipeline runs via `GET /projects/:id/pipelines`
4. Add the scheduled pipeline descriptions to the relevant pipeline runs
5. Filter the pipelines

**Comment by javanegmond** [comment]
That seems a bit excessive, doesn't it? Instead of getting all pipelines via GET `/projects/:id/pipelines` and enriching them with pipeline schedule data it's also possible to obtain all pipelines attached to a schedule via `GET /projects/:id/pipeline_schedules/:pipeline_schedule_id/pipelines`

https://docs.gitlab.com/ee/api/pipeline_schedules.html#get-all-pipelines-triggered-by-a-pipeline-schedule

**Comment by fniessink** [comment]
I don't see what's excessive. The user may want to see scheduled and non-scheduled pipelines, so we can't always get only the scheduled pipelines. Also, if multiple schedules match the given pipeline description, we still need multiple calls to `GET /projects/:id/pipeline_schedules/:pipeline_schedule_id/pipelines` (one for each matched schedule). 

I guess step 4 in the flow sketched above can be prevented by doing the filtering by pipeline schedule description earlier, but it's not like the flow becomes much simpler:

1. If the Quality-time parameter `pipeline_triggers_to_include` includes `schedule`:
    1.  Get all pipeline schedules and descriptions via `GET /projects/:id/pipeline_schedules`
    2. Filter the pipeline schedules by pipeline schedule description
    3. For each filtered pipeline schedule, get the pipeline runs via `GET /projects/:id/pipeline_schedules/:pipeline_schedule_id/pipelines`
1. If the Quality-time parameter `pipeline_triggers_to_include` includes other triggers besides `schedule`:
    1. Get pipeline runs that don't have source `schedule` via `GET /projects/:id/pipelines`
1. Combine the pipeline runs from step 1 and 2
1. Apply any other filters (branches, status)

Notes to self:
- `GET /projects/:id/pipelines` allows for passing `status` as query parameter, but `GET /projects/:id/pipeline_schedules/:pipeline_schedule_id/pipelines` does not.
- If the user wants to filter by pipeline schedule description but doesn't include the trigger `schedule` in the `pipeline_triggers_to_include` parameter value, no scheduled pipelines are included. Add a warning to the parameter.
- The measurement entities should show the matched pipelines
